compiler/rustc_infer/src/infer/relate/combine.rs

```diff
@@ -180,30 +180,14 @@ impl<'tcx> InferCtxt<'tcx> {
                 &mut OriginalQueryValues::default(),
             );
             self.tcx.check_tys_might_be_eq(canonical).map_err(|_| {
+                // The error will only be reported later. If we emit an ErrorGuaranteed
+                // here, then we will never get to the code that actually emits the error.
                 self.tcx.dcx().delayed_bug(format!(
                     "cannot relate consts of different types (a={a:?}, b={b:?})",
-                ))
+                ));
+                TypeError::Mismatch
             })
-        });
-
-        // If the consts have differing types, just bail with a const error with
-        // the expected const's type. Specifically, we don't want const infer vars
-        // to do any type shapeshifting before and after resolution.
-        if let Err(guar) = compatible_types {
-            // HACK: equating both sides with `[const error]` eagerly prevents us
-            // from leaving unconstrained inference vars during things like impl
-            // matching in the solver.
-            let a_error = ty::Const::new_error(self.tcx, guar, a.ty());
-            if let ty::ConstKind::Infer(InferConst::Var(vid)) = a.kind() {
-                return self.unify_const_variable(vid, a_error, relation.param_env());
-            }
-            let b_error = ty::Const::new_error(self.tcx, guar, b.ty());
-            if let ty::ConstKind::Infer(InferConst::Var(vid)) = b.kind() {
-                return self.unify_const_variable(vid, b_error, relation.param_env());
-            }
-
-            return Ok(if relation.a_is_expected() { a_error } else { b_error });
-        }
+        })?;
```

Comment:
This is an example where we create a delayed bug, but never actually report an error. The error is reported later, but we don't get there anymore, because the `ty::Const::new_error` we created here make us bail out early right after coherence checking.

Reply:
I pulled out this change in isolation: https://github.com/rust-lang/rust/pull/120059 it's not great, but unsure how to handle this better. We have the follwing options:

* do what we do now, bubbling up a `ConstKind::Error`, causes follow-up issues, because other code assumes errors have already been reported
* return a `TypeError::Mismatch`: means that impls with errors won't be found, causing follow-up errors about types not implementing traits that obviously do, but with errors
* just go through the happy path as if the types matched, later code will error out. Weird, and produces odd diagnostics: https://github.com/rust-lang/rust/pull/120059